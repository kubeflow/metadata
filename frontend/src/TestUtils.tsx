/*
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* eslint-disable */
// tslint:disable: object-literal-sort-keys

// Because this is test utils.

import * as React from 'react';
// @ts-ignore
import createRouterContext from 'react-router-test-context';
import {PageProps, Page} from './pages/Page';
import {ToolbarActionConfig} from './components/Toolbar';
import {match} from 'react-router';
import {mount, ReactWrapper} from 'enzyme';
import {object} from 'prop-types';
import {format} from 'prettier';
import snapshotDiff from 'snapshot-diff';
import {
  Artifact,
  ArtifactProperties,
  ArtifactCustomProperties,
  Value,
} from '@kubeflow/frontend';

export default class TestUtils {
  /**
   * Mounts the given component with a fake router and returns the mounted tree
   */
  // tslint:disable-next-line:variable-name
  public static mountWithRouter(component: React.ReactElement<any>): ReactWrapper {
    const childContextTypes = {
      router: object,
    };
    const context = createRouterContext();
    const tree = mount(component, { context, childContextTypes });
    return tree;
  }

  /**
   * Flushes all already queued promises and returns a promise. Note this will
   * only work if the promises have already been queued, so it cannot be used to
   * wait on a promise that hasn't been dispatched yet.
   */
  public static flushPromises(): Promise<void> {
    return new Promise(resolve => setImmediate(resolve));
  }

  /**
   * Adds a one-time mock implementation to the provided spy that mimics an error
   * network response
   */
  public static makeErrorResponseOnce(spy: jest.MockInstance<unknown, unknown[]>, message: string): void {
    spy.mockImplementationOnce(() => {
      throw {
        text: () => Promise.resolve(message),
      };
    });
  }

  /**
   * Generates a customizable PageProps object that can be passed to initialize
   * Page components, taking care of setting ToolbarProps properly, which have
   * to be set after component initialization.
   */
  // tslint:disable-next-line:variable-name
  public static generatePageProps(
    PageElement: new (_: PageProps) => Page<any, any>,
    location: Location,
    matchValue: match,
    historyPushSpy: jest.SpyInstance<unknown> | null,
    updateBannerSpy: jest.SpyInstance<unknown> | null,
    updateDialogSpy: jest.SpyInstance<unknown> | null,
    updateToolbarSpy: jest.SpyInstance<unknown> | null,
    updateSnackbarSpy: jest.SpyInstance<unknown> | null,
  ): PageProps {
    const pageProps = {
      history: { push: historyPushSpy } as any,
      location: location as any,
      match: matchValue,
      toolbarProps: { actions: {}, breadcrumbs: [], pageTitle: '' },
      updateBanner: updateBannerSpy as any,
      updateDialog: updateDialogSpy as any,
      updateSnackbar: updateSnackbarSpy as any,
      updateToolbar: updateToolbarSpy as any,
    } as PageProps;
    pageProps.toolbarProps = new PageElement(pageProps).getInitialToolbarState();
    // The toolbar spy gets called in the getInitialToolbarState method, reset it
    // in order to simplify tests
    if (updateToolbarSpy) {
      updateToolbarSpy.mockReset();
    }
    return pageProps;
  }

  public static getToolbarButton(
    updateToolbarSpy: jest.SpyInstance<unknown>,
    buttonKey: string,
  ): ToolbarActionConfig {
    const lastCallIdx = updateToolbarSpy.mock.calls.length - 1;
    const lastCall = updateToolbarSpy.mock.calls[lastCallIdx][0];
    return lastCall.actions[buttonKey];
  }
}

/**
 * Generate diff text for two HTML strings.
 * Recommend providing base and update annotations to clarify context in the diff directly.
 */
export function diffHTML({
  base,
  update,
  baseAnnotation,
  updateAnnotation,
}: {
  base: string;
  baseAnnotation?: string;
  update: string;
  updateAnnotation?: string;
}) {
  return diff({
    base: formatHTML(base),
    update: formatHTML(update),
    baseAnnotation,
    updateAnnotation,
  });
}

export function diff({
  base,
  update,
  baseAnnotation,
  updateAnnotation,
}: {
  base: string;
  baseAnnotation?: string;
  update: string;
  updateAnnotation?: string;
}) {
  return snapshotDiff(base, update, {
    stablePatchmarks: true, // Avoid line numbers in diff, so that diffs are stable against irrelevant changes
    aAnnotation: baseAnnotation,
    bAnnotation: updateAnnotation,
  });
}

function formatHTML(html: string): string {
  return format(html, { parser: 'html' });
}

/**	
 * Generates a customizable PageProps object that can be passed to initialize	
 * Page components, taking care of setting ToolbarProps properly, which have	
 * to be set after component initialization.	
 */	
// tslint:disable-next-line:variable-name	
export function generatePageProps(	
  // tslint:disable-next-line: variable-name
  PageElement: new (_: PageProps) => Page<any, any>, location: Location,	
  matchValue: match, historyPushSpy: jest.SpyInstance | null,	
  updateBannerSpy: jest.SpyInstance | null,	
  updateDialogSpy: jest.SpyInstance | null,	
  updateToolbarSpy: jest.SpyInstance | null,	
  updateSnackbarSpy: jest.SpyInstance | null): PageProps {	
  const pageProps = {	
    history: {push: historyPushSpy} as any,	
    location: location as any,	
    match: matchValue,	
    toolbarProps: {actions: [], breadcrumbs: [], pageTitle: ''},	
    updateBanner: updateBannerSpy as any,	
    updateDialog: updateDialogSpy as any,	
    updateSnackbar: updateSnackbarSpy as any,	
    updateToolbar: updateToolbarSpy as any,	
  } as unknown as PageProps;	
  pageProps.toolbarProps = new PageElement(pageProps).getInitialToolbarState();	
  // The toolbar spy gets called in the getInitialToolbarState method, reset	
  // it in order to simplify tests	
  if (updateToolbarSpy) {	
    updateToolbarSpy.mockReset();	
  }	
  return pageProps;	
}	

export function getToolbarButton(	
  updateToolbarSpy: jest.SpyInstance, title: string): ToolbarActionConfig {	
  const lastCallIdx = updateToolbarSpy.mock.calls.length - 1;	
  const lastCall = updateToolbarSpy.mock.calls[lastCallIdx][0];	
  return lastCall.actions.find((b: any) => b.title === title);	
}	

export const doubleValue = (number: number) => {	
  const value = new Value();	
  value.setDoubleValue(number);	
  return value;	
};	

export const intValue = (number: number) => {	
  const value = new Value();	
  value.setIntValue(number);	
  return value;	
};	

export const stringValue = (string: String) => {	
  const value = new Value();	
  value.setStringValue(String(string));	
  return value;	
};	

export const buildTestModel = () => {	
  const model = new Artifact();	
  model.setId(1);	
  model.setTypeId(1);	
  model.setUri('gs://my-bucket/mnist');	
  model.getPropertiesMap().set(ArtifactProperties.NAME, stringValue('test model'));	
  model.getPropertiesMap().set(ArtifactProperties.DESCRIPTION, stringValue('A really great model'));	
  model.getPropertiesMap().set(ArtifactProperties.VERSION, stringValue('v1'));	
  model.getPropertiesMap().set(ArtifactProperties.CREATE_TIME, stringValue('2019-06-12T01:21:48.259263Z'));	
  model.getPropertiesMap().set(ArtifactProperties.ALL_META, stringValue(	
      '{"hyperparameters": {"early_stop": true, ' +	
      '"layers": [10, 3, 1], "learning_rate": 0.5}, ' +	
      '"model_type": "neural network", ' +	
      '"training_framework": {"name": "tensorflow", "version": "v1.0"}}'));	
  model.getCustomPropertiesMap().set(ArtifactCustomProperties.WORKSPACE, stringValue('workspace-1'));	
  model.getCustomPropertiesMap().set(ArtifactCustomProperties.RUN, stringValue('1'));	
  return model	
};	

export const testModel = buildTestModel();