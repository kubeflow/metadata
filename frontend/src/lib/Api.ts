import {MetadataServiceApi} from '../apis/service';
import {
  MetadataStoreServiceClient,
  ServiceError, UnaryResponse
} from "../generated/src/apis/metadata/metadata_store_service_pb_service";

/** Known Artifact properties */
export enum ArtifactProperties {
  ALL_META = '__ALL_META__',
  CREATE_TIME = 'create_time',
  DESCRIPTION = 'description',
  NAME = 'name',
  PIPELINE_NAME = 'pipeline_name',
  VERSION = 'version',
}

/** Known Artifact custom properties */
export enum ArtifactCustomProperties {
  WORKSPACE = '__kf_workspace__',
  RUN = '__kf_run__',
}

/** Known Execution properties */
export enum ExecutionProperties {
  NAME = 'name',
  PIPELINE_NAME = 'pipeline_name',
  STATE = 'state',
}

/** Known Execution custom properties */
export enum ExecutionCustomProperties {
  WORKSPACE = '__kf_workspace__',
}

/** Format for a List request */
export interface ListRequest {
  filter?: string;
  orderAscending?: boolean;
  pageSize?: number;
  pageToken?: string;
  sortBy?: string;
}

type Callback<R> = (err: ServiceError | null, res: R | null) => void;
type MetadataApiMethod<T, R> = (request: T, callback: Callback<R>) => UnaryResponse;
type PromiseBasedMetadataApiMethod<T, R> = (request: T) => Promise<{ response: R | null, error: ServiceError | null }>;

/**
 * Converts a callback based api method to promise based api method.
 */
function makePromiseApi<T, R>(apiMethod: MetadataApiMethod<T, R>): PromiseBasedMetadataApiMethod<T, R> {
  return (request: T) => new Promise((resolve, reject) => {
    const handler = (error: ServiceError | null, response: R | null) => {
      // resolve both response and error to keep type information
      resolve({ response, error });
    };
    apiMethod(request, handler);
  });
}

const metadataServiceClient = new MetadataStoreServiceClient('');

// TODO: add all other api methods we need here.
const metadataServicePromiseClient = {
  getArtifactTypes: makePromiseApi(metadataServiceClient.getArtifactTypes.bind(metadataServiceClient)),
  getArtifactsByID: makePromiseApi(metadataServiceClient.getArtifactsByID.bind(metadataServiceClient)),
  getEventsByArtifactIDs: makePromiseApi(metadataServiceClient.getEventsByArtifactIDs.bind(metadataServiceClient)),
  getEventsByExecutionIDs: makePromiseApi(metadataServiceClient.getEventsByExecutionIDs.bind(metadataServiceClient)),
  getExecutionsByID: makePromiseApi(metadataServiceClient.getExecutionsByID.bind(metadataServiceClient)),
};

/**
 * Class to wrap backend APIs.
 */
export class Api {

  private static instance: Api;
  private metadataServiceApi: MetadataServiceApi;

  /**
   * Factory function to return an Api instance.
   */
  public static getInstance(): Api {
    if (!Api.instance) {
      const location = window.location;
      let path = `${location.protocol}//${location.host}${location.pathname}`;
      if (path.endsWith('/')) path = path.slice(0, path.length - 1);
      Api.instance = new Api(path);
    }
    return Api.instance;
  }

  private constructor(basePath: string) {
    this.metadataServiceApi = new MetadataServiceApi({basePath});
  }

  get metadataService(): MetadataServiceApi {
    return this.metadataServiceApi;
  }

  get metadataStoreService() {
    return metadataServicePromiseClient;
  }
}
