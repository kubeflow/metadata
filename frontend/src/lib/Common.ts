/** Format for a List request */
export interface ListRequest {
  filter?: string;
  orderAscending?: boolean;
  pageSize?: number;
  pageToken?: string;
  sortBy?: string;
}
