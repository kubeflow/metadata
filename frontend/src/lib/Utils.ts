/*
 * Copyright 2019 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {isFunction} from 'lodash';
import {MlMetadataArtifact, MlMetadataValue} from '../apis/service';

export const logger = {
  error: (...args: any[]) => {
    // tslint:disable-next-line:no-console
    console.error(...args);
  },
  verbose: (...args: any[]) => {
    // tslint:disable-next-line:no-console
    console.log(...args);
  },
};

export function formatDateString(date: Date | string | undefined): string {
  if (typeof date === 'string') {
    return new Date(date).toLocaleString();
  } else {
    return date ? date.toLocaleString() : '-';
  }
}

export async function errorToMessage(error: any): Promise<string> {
  if (error instanceof Error) {
    return error.message;
  }

  if (error && error.text && isFunction(error.text)) {
    return await error.text();
  }

  return JSON.stringify(error) || '';
}

/** Title cases a string by capitalizing the first letter of each word. */
export function titleCase(str: string): string {
  return str.split(/[\s_-]/)
    .map((w) => `${w.charAt(0).toUpperCase()}${w.slice(1)}`)
    .join(' ');
}

/**
 * Safely extracts the named property or custom property from the provided
 * Artifact.
 * @param artifact
 * @param propertyName
 * @param fromCustomProperties
 */
export function getArtifactProperty(artifact: MlMetadataArtifact,
  propertyName: string, fromCustomProperties = false): string | null {
  const props = fromCustomProperties ?
    artifact.custom_properties : artifact.properties;

  return (props && props[propertyName] && props[propertyName].string_value)
    || null;
}

export function getMetadataValue(mlMetadataValue: MlMetadataValue): string | number {
  return mlMetadataValue.double_value || mlMetadataValue.int_value || mlMetadataValue.string_value || '';
}
