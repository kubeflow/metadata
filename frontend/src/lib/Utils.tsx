/*
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import {isFunction} from 'lodash';
import {ListRequest} from '@kubeflow/frontend';
import {Row, Column, ExpandState, css} from '../components/CustomTable';
import {padding} from '../Css';
import {classes} from 'typestyle';
import {CustomTableRow} from '../components/CustomTableRow';

export const logger = {
  error: (...args: any[]) => {
    // tslint:disable-next-line:no-console
    console.error(...args);
  },
  warn: (...args: any[]) => {
    // tslint:disable-next-line:no-console
    console.warn(...args);
  },
  verbose: (...args: any[]) => {
    // tslint:disable-next-line:no-console
    console.log(...args);
  },
};

export function extendError(err: any, extraMessage?: string): any {
  if (err.message && typeof err.message === 'string') {
    err.message = extraMessage + ': ' + err.message;
  }
  return err;
}

export function rethrow(err: any, extraMessage?: string): never {
  throw extendError(err, extraMessage);
}

export function formatDateString(date: Date | string | undefined): string {
  if (typeof date === 'string') {
    return new Date(date).toLocaleString();
  } else {
    return date ? date.toLocaleString() : '-';
  }
}

// TODO: add tests
export async function errorToMessage(error: any): Promise<string> {
  if (error instanceof Error) {
    return error.message;
  }

  if (error && error.text && isFunction(error.text)) {
    return await error.text();
  }

  return JSON.stringify(error) || '';
}

export function s(items: any[] | number): string {
  const length = Array.isArray(items) ? items.length : items;
  return length === 1 ? '' : 's';
}

interface ServiceError {
  message: string;
  code?: number;
}

export function serviceErrorToString(error: ServiceError): string {
  return `Error: ${error.message}.${error.code ? ` Code: ${error.code}` : ''}`;
}

/**
 * Returns true if no filter is specified, or if the filter string matches any of the row's columns,
 * case insensitively.
 * @param request
 */
export function rowFilterFn(request: ListRequest): (r: Row) => boolean {
  // TODO: We are currently searching across all properties of all artifacts. We should figure
  // what the most useful fields are and limit filtering to those
  return r => {
    if (!request.filter || /^[a-z]+$/.test(request.filter)) {
      return true;
    }

    const decodedFilter = decodeURIComponent(request.filter);
    try {      
      const filter = JSON.parse(decodedFilter);
      if (!filter.predicates || filter.predicates.length === 0) {
        return true;
      }
      // TODO: Extend this to look at more than a single predicate
      const filterString =
        '' +
        (filter.predicates[0].int_value ||
          filter.predicates[0].long_value ||
          filter.predicates[0].string_value);
      return (
        r.otherFields
          .join('')
          .toLowerCase()
          .indexOf(filterString.toLowerCase()) > -1
      );
    } catch (err) {
      logger.error('Error parsing request filter!', err);
      console.log(`[${decodedFilter}], [${request.filter}]`)
      return true;
    }
  };
}

export function rowCompareFn(
  request: ListRequest,
  columns: Column[],
): (r1: Row, r2: Row) => number {
  return (r1, r2) => {
    if (!request.sortBy) {
      return -1;
    }

    const descSuffix = ' desc';
    const cleanedSortBy = request.sortBy.endsWith(descSuffix)
      ? request.sortBy.substring(0, request.sortBy.length - descSuffix.length)
      : request.sortBy;

    const sortIndex = columns.findIndex(c => cleanedSortBy === c.sortKey);

    // Convert null to string to avoid null comparison behavior
    const compare = (r1.otherFields[sortIndex] || '') < (r2.otherFields[sortIndex] || '');
    if (request.orderAscending) {
      return compare ? -1 : 1;
    } else {
      return compare ? 1 : -1;
    }
  };
}

export interface CollapsedAndExpandedRows {
  // collapsedRows are the first row of each group, what the user sees before expanding a group.
  collapsedRows: Row[];
  // expandedRows is a map of grouping keys to a list of rows grouped by that key. This is what a
  // user sees when they expand one or more rows.
  expandedRows: Map<number, Row[]>;
}

/**
 * Groups the incoming rows by name and type pushing all but the first row
 * of each group to the expandedRows Map.
 * @param rows
 */
export function groupRows(rows: Row[]): CollapsedAndExpandedRows {
  const flattenedRows = rows.reduce((map, r) => {
    const stringKey = r.otherFields[0];
    const rowsForKey = map.get(stringKey);
    if (rowsForKey) {
      rowsForKey.push(r);
    } else {
      map.set(stringKey, [r]);
    }
    return map;
  }, new Map<string, Row[]>());

  const collapsedAndExpandedRows: CollapsedAndExpandedRows = {
    collapsedRows: [],
    expandedRows: new Map<number, Row[]>(),
  };
  // collapsedRows are the first row of each group, what the user sees before expanding a group.
  Array.from(flattenedRows.entries()) // entries() returns in insertion order
    .forEach((entry, index) => {
      // entry[0] is a grouping key, entry[1] is a list of rows
      const rowsInGroup = entry[1];

      // If there is only one row in the group, don't allow expansion.
      // Only the first row is displayed when collapsed
      if (rowsInGroup.length === 1) {
        rowsInGroup[0].expandState = ExpandState.NONE;
      }

      // Add the first row in this group to be displayed as collapsed row
      collapsedAndExpandedRows.collapsedRows.push(rowsInGroup[0]);

      // Remove the grouping column text for all but the first row in the group because it will be
      // redundant within an expanded group.
      const hiddenRows = rowsInGroup.slice(1);
      hiddenRows.forEach(row => (row.otherFields[0] = ''));

      // Add this group of rows sharing a pipeline to the list of grouped rows
      collapsedAndExpandedRows.expandedRows.set(index, hiddenRows);
    });

  return collapsedAndExpandedRows;
}

/**
 * Returns a fragment representing the expanded content for the given
 * row.
 * @param index
 */
export function getExpandedRow(
  expandedRows: Map<number, Row[]>,
  columns: Column[],
): (index: number) => React.ReactNode {
  return (index: number) => {
    const rows = expandedRows.get(index) || [];

    return (
      <div className={padding(65, 'l')}>
        {rows.map((r, rindex) => (
          <div className={classes('tableRow', css.row)} key={rindex}>
            <CustomTableRow row={r} columns={columns} />
          </div>
        ))}
      </div>
    );
  };
}
