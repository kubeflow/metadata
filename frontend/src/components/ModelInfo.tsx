/*
 * Copyright 2019 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import * as React from 'react';
import {stylesheet, classes} from 'typestyle';
import {color, commonCss} from '../Css';
import {formatDateString} from '../lib/Utils';

const css = stylesheet({
  modelInfo: {
    display: 'flex',
    flexDirection: 'row',
    flexWrap: 'wrap',
  },
  field: {
    flexBasis: '300px',
    marginBottom: '32px',
  },
  term: {
    color: color.grey,
    fontSize: '12px',
    letterSpacing: '0.2px',
    lineHeight: '16px',
  },
  value: {
    color: color.secondaryText,
    fontSize: '14px',
    letterSpacing: '0.2px',
    lineHeight: '20px',
  }
});

// Standard artifact properties and Model properties from
// https://github.com/kubeflow/metadata/blob/master/schema/alpha/artifacts/model.json
export interface ModelInfoProps {
  description?: string;
  version?: string;
  createTime?: string;
  uri?: string;
  workspace?: string;
  run?: string;
  apiVersion?: string;
  hyperparameters?: {[name: string]: {}};
  modelType?: string;
  trainingFramework?: {name: string, version: string};
}

export class ModelInfo extends React.Component<ModelInfoProps, {}> {

  public render(): JSX.Element {
    const createTime = formatDateString(this.props.createTime);
    return (
      <section>
        <h1 className={commonCss.header}>Model info</h1>
        <dl className={css.modelInfo}>
          <div className={css.field}>
            <dt className={css.term}>Type</dt>
            <dd className={css.value}>{this.props.modelType}</dd>
          </div>
          <div className={css.field}>
            <dt className={css.term}>Description</dt>
            <dd className={css.value}>{this.props.description}</dd>
          </div>
          <div className={css.field}>
            <dt className={css.term}>Version ID</dt>
            <dd className={css.value}>{this.props.version}</dd>
          </div>
          <div className={css.field}>
            <dt className={css.term}>Workspace</dt>
            <dd className={css.value}>{this.props.workspace}</dd>
          </div>
          <div className={css.field}>
            <dt className={css.term}>Run</dt>
            <dd className={css.value}>{this.props.run}</dd>
          </div>
          <div className={css.field}>
            <dt className={css.term}>Creation time</dt>
            <dd className={css.value}>{createTime}</dd>
          </div>
          <div className={css.field}>
            <dt className={css.term}>Hyperparameters</dt>
            <dd className={classes(css.value, commonCss.flexColumn)}>{
              this.props.hyperparameters &&
              Object.entries(this.props.hyperparameters)
                .map((e, i) => <span key={i}>{`${e[0]}: ${e[1]}`}</span>)
            }
            </dd>
          </div>
          <div className={css.field}>
            <dt className={css.term}>Training data</dt>
            <dd className={css.value}>{this.props.uri}</dd>
          </div>
          <div className={css.field}>
            <dt className={css.term}>Training framework</dt>
            <dd className={css.value}>
              {
                this.props.trainingFramework ?
                  `${this.props.trainingFramework.name}
                   ${this.props.trainingFramework.version || ''}` : ''
              }
            </dd>
          </div>
          <div className={css.field}>
            <dt className={css.term}>Training data</dt>
            <dd className={css.value}>{this.props.uri}</dd>
          </div>
        </dl>
      </section>
    );
  }
}
