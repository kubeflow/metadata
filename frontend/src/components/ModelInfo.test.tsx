import * as React from 'react';
import {ModelInfo} from './ModelInfo';
import {shallow, ShallowWrapper} from 'enzyme';
import {buildTestModel, stringValue, testModel} from '../TestUtils';
import {ArtifactProperties} from '../lib/Api';

describe('ModelInfo', () => {
  let tree: ShallowWrapper;

  const consoleErrorSpy = jest.spyOn(console, 'error')
    .mockImplementation(() => null);

  afterEach(() => tree.unmount());

  it('Renders Model information', () => {
    tree = shallow(<ModelInfo model={testModel} />);

    expect(tree).toMatchSnapshot();
  });

  it('Renders Model information with missing __ALL_META__ property', () => {
    const model = buildTestModel();
    model.getPropertiesMap().del(ArtifactProperties.ALL_META);

    tree = shallow(<ModelInfo model={model} />);

    expect(tree).toMatchSnapshot();
  });

  it('Renders Model information with malformed __ALL_META__ property', () => {
    const model = buildTestModel();
    model.getPropertiesMap().set(ArtifactProperties.ALL_META, stringValue('non-JSON string'));

    tree = shallow(<ModelInfo model={model} />);

    expect(tree).toMatchSnapshot();
    expect(consoleErrorSpy.mock.calls[0][0]).toBe(
      'Unable to parse __ALL_META__ property');
  });
});
