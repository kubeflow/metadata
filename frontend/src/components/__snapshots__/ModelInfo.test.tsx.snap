// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModelInfo Renders Model information 1`] = `
<section>
  <h1
    className="header_f189n3y"
  >
    Model info
  </h1>
  <dl
    className="modelInfo_fnuket9"
  >
    <div
      className="field_f1bgv1eo"
    >
      <dt
        className="term_fuqr2el"
      >
        Type
      </dt>
      <dd
        className="value_f1d9lp3k"
      >
        neural network
      </dd>
    </div>
    <div
      className="field_f1bgv1eo"
    >
      <dt
        className="term_fuqr2el"
      >
        Description
      </dt>
      <dd
        className="value_f1d9lp3k"
      >
        model to recognize handwritten digits
      </dd>
    </div>
    <div
      className="field_f1bgv1eo"
    >
      <dt
        className="term_fuqr2el"
      >
        Version ID
      </dt>
      <dd
        className="value_f1d9lp3k"
      >
        v0.0.1
      </dd>
    </div>
    <div
      className="field_f1bgv1eo"
    >
      <dt
        className="term_fuqr2el"
      >
        Workspace
      </dt>
      <dd
        className="value_f1d9lp3k"
      >
        workspace1
      </dd>
    </div>
    <div
      className="field_f1bgv1eo"
    >
      <dt
        className="term_fuqr2el"
      >
        Run
      </dt>
      <dd
        className="value_f1d9lp3k"
      >
        run123
      </dd>
    </div>
    <div
      className="field_f1bgv1eo"
    >
      <dt
        className="term_fuqr2el"
      >
        Creation time
      </dt>
      <dd
        className="value_f1d9lp3k"
      >
        6/11/2019, 9:21:48 PM
      </dd>
    </div>
    <div
      className="field_f1bgv1eo"
    >
      <dt
        className="term_fuqr2el"
      >
        Hyperparameters
      </dt>
      <dd
        className="value_f1d9lp3k flexColumn_fr3ikkd"
      >
        <span
          key="0"
        >
          early_stop: true
        </span>
        <span
          key="1"
        >
          layers: 10,3,1
        </span>
        <span
          key="2"
        >
          learning_rate: 0.5
        </span>
      </dd>
    </div>
    <div
      className="field_f1bgv1eo"
    >
      <dt
        className="term_fuqr2el"
      >
        Training data
      </dt>
      <dd
        className="value_f1d9lp3k"
      >
        gs://my-bucket/mnist
      </dd>
    </div>
    <div
      className="field_f1bgv1eo"
    >
      <dt
        className="term_fuqr2el"
      >
        Training framework
      </dt>
      <dd
        className="value_f1d9lp3k"
      >
        tensorflow
                   v1.0
      </dd>
    </div>
    <div
      className="field_f1bgv1eo"
    >
      <dt
        className="term_fuqr2el"
      >
        Training data
      </dt>
      <dd
        className="value_f1d9lp3k"
      >
        gs://my-bucket/mnist
      </dd>
    </div>
  </dl>
</section>
`;
